Singly linked list

zfletch opened the pull request:
Add a singly linked list. I think it makes sense to have different types of linked lists in this application, so I wrote an implementation of a singly linked one (to compare to the doubly linked list that already exists).

``` ruby
sl = SinglyLinkedList.new # => <>
dl = DoublyLinkedList.new # => <<>>

sl.push(1,2,3,4) # => <1, 2, 3, 4>
dl.push(5,6,7,8) # => <<5, 6, 7, 8>>

sl.pop # => 4
dl.pop # => 8

sl # => <1, 2, 3>
dl # => <<5, 6, 7>>
```

dgoldhirsch commented on app/models/singly_linked_list.rb:
```diff
@@ -0,0 +1,77 @@
+class SinglyLinkedList
+  include Enumerable
```
Enumerable is a very large module.  The reason I avoided including it was that I didn't want to have to assert that things like Enumerable#grep and #include? and #inject? and #map worked.  But, perhaps, they do work if we provide the basic #each?  That'd be very, very slick.  If we don't know that they work, then, I'd suggest not including the module (unless you also provide test-coverage for it).  But, please, what do you think?

zfletch replied:
I agree that we shouldn't test `#grep`, `#include?`, etc. But the `Enumerable` module is basically designed to work correctly as long as there is an `#each` method defined. From http://ruby-doc.org/core-2.3.0/Enumerable.html:

> The Enumerable mixin provides collection classes with several traversal and searching methods, and with the ability to sort. The class must provide a method each, which yields successive members of the collection.

I think we can trust Enumerable as long as `each` yields each member of the list in turn, it's pretty robust. In fact, it even works with an infinite `each` method. For example:

``` ruby
class Fib
  include Enumerable

  def each
    first, second = 0, 1
    loop do
      yield second
      first, second = second, first + second
    end
  end
end

Fib.new.first(10) # => [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]
```

dgoldhirsch replied:
That's way cool!  Well, I suppose...

I'm rusty on infinite closures....I would have thought that a break or a
return were necessary somewhere in there.  Does the yield automatically
break out?  That seems wild, to me.

On Mon, Feb 1, 2016 at 5:53 PM, zfletch notifications@github.com wrote:

> In app/models/singly_linked_list.rb
> https://github.com/dgoldhirsch/little-stuff/pull/3#discussion_r51495039:
> 
> > @@ -0,0 +1,77 @@
> > +class SinglyLinkedList
> > -  include Enumerable
> 
> I agree that we shouldn't test #grep, #include?, etc. But the Enumerable
> module is basically designed to work correctly as long as there is an
> #each method defined. From http://ruby-doc.org/core-2.3.0/Enumerable.html:
> 
> The Enumerable mixin provides collection classes with several traversal
> and searching methods, and with the ability to sort. The class must provide
> a method each, which yields successive members of the collection.
> 
> I think we can trust Enumerable as long as each yields each member of the
> list in turn, it's pretty robust. In fact, it even works with an infinite
> each method. For example:
> 
> class Fib
>   include Enumerable
> 
>   def each
>     first, second = 0, 1
>     loop do
>       yield second
>       first, second = second, first + second
>     end
>   endend
> Fib.new.first(10) # => [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dgoldhirsch/little-stuff/pull/3/files#r51495039.

zfletch replied:
The code that does this is written in C (I think it's [this function](https://github.com/ruby/ruby/blob/6ee8ec70696260fb0ac2ab4bf13f40ff003a8995/vm_eval.c#L1147)), but you can get the same behavior with Ruby code that looks like:

``` ruby
def first(max)
  count = 0
  acc = []

  each do |n|
    break if count == max
    acc << n
    count += 1
  end

  acc
end
```

dgoldhirsch replied:
Yeah, that looks more familiar.  Hey, I'm trying (and failing) to connect
with you via my Adium IM client.
Are you logged in to Google at the moment?

I'm thinking, let's merge your PRs into little-stuff.  Then, if you're
willing/interested, please tell me what you think about my slightly
different RSpec test for my linked list?

I'm going over to ezcater this afternoon with my linked list (they've
already seen the original version of this) and the background enroller and
its tests/features, and some scheduling stuff I did back at the DMC
(previous job).

What'cha up to?  Having single and double-linked lists seems like a very
good idea, along with all kinds of other CS kata.

On Tue, Feb 2, 2016 at 10:27 AM, zfletch notifications@github.com wrote:

> In app/models/singly_linked_list.rb
> https://github.com/dgoldhirsch/little-stuff/pull/3#discussion_r51582881:
> 
> > @@ -0,0 +1,77 @@
> > +class SinglyLinkedList
> > -  include Enumerable
> 
> The code that does this is written in C (I think it's this function
> https://github.com/ruby/ruby/blob/6ee8ec70696260fb0ac2ab4bf13f40ff003a8995/vm_eval.c#L1147),
> but you can get the same behavior with Ruby code that looks like:
> 
> def first(max)
>   count = 0
>   acc = []
> 
>   each do |n|
>     break if count == max
>     acc << n
>     count += 1
>   end
> 
>   accend
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dgoldhirsch/little-stuff/pull/3/files#r51582881.

dgoldhirsch commented on app/models/doubly_linked_list.rb:
```diff
@@ -29,12 +31,16 @@ def each(&block)
-      self.head = self.tail = Node.new(value)
-    else
-      self.tail = self.tail.right = Node.new(value, tail)
+  def push(*values)
```
I thought about this, but decided that I had not been asked to provide a multi-value #push.  #push (or #<<) is not defined by Enumerable, in Ruby.  It appears to be defined in Array.  But, there, it is defined specifically for one and only one parameter value (I have wished it to take a splat-list, many times in the past).

That all said, I can't think of a good reason not to support a splat-list.  It'd be awfully convenient.

Well, one thing....Array#<< returns the object being pushed into the array.  I don't like that--I think it'd be better if it returned the receiver, so that one could write "array << 1 << 10 << 20...", etc.  But, that's what Array#<< does.  I suppose there's no good reason (other than consistency of thought) why this little class has to be consistent with Array, however.

zfletch replied:
Yeah, push is defined in `Array`, but I believe that it can take zero arguments or multiple arguments. (http://ruby-doc.org/core-2.3.0/Array.html#method-i-push).

I also agree that it should return the array and not the pushed value (which is why I added the `self` at the end here). I think `<<` is implemented differently than `push` and I haven't done anything with that method.

I think the implementation in this PR matches the implementation in `Array`.

zfletch committed:
singly linked list

Add a singly linked list. I think it makes sense to have different types
of linked lists in this application, so I wrote an implementation of a
singly linked one (to compare to the doubly linked list that already
exists). I also renamed the doubly linked list.
